File: android-beans/src/test/java/rocks/frieler/android/beans/scopes/singleton/SingletonScopedFactoryBeanHandlerTest.java
package rocks.frieler.android.beans.scopes.singleton;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.sameInstance;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static rocks.frieler.android.beans.scopes.singleton.SingletonScopedFactoryBeanHandler.SINGLETON_SCOPE;

@RunWith(MockitoJUnitRunner.class)
public class SingletonScopedFactoryBeanHandlerTest {
    private SingletonScopedFactoryBeanHandler singletonScopedFactoryBeanHandler = new SingletonScopedFactoryBeanHandler();

    @Mock
    private SingletonScopedFactoryBean<Object> scopedFactoryBean;

    @Test
    public void testGetNameReturnsPrototypeScope() {
        assertThat(singletonScopedFactoryBeanHandler.getName(), is(SINGLETON_SCOPE));
    }

    @Test
    public void testIsAlwaysActive() {
        assertThat(singletonScopedFactoryBeanHandler.isActive(), is(true));
    }

    @Test
    public void testGetBeanLetsFactoryBeanProduceBean() {
        final Object bean = new Object();
        when(scopedFactoryBean.produceBean()).thenReturn(bean);

        final Object beanInstance = singletonScopedFactoryBeanHandler.getBean("bean", scopedFactoryBean);

        assertThat(beanInstance, is(sameInstance(bean)));
    }

    @Test
    public void testGetBeanProducesNewBeanEveryTime() {
        final Object bean = new Object();
        when(scopedFactoryBean.produceBean()).thenReturn(bean);

        final Object firstBeanInstance = singletonScopedFactoryBeanHandler.getBean("bean", scopedFactoryBean);
        final Object secondBeanInstance = singletonScopedFactoryBeanHandler.getBean("bean", scopedFactoryBean);

        verify(scopedFactoryBean, times(1)).produceBean();
        assertThat(firstBeanInstance, is(sameInstance(bean)));
        assertThat(secondBeanInstance, is(sameInstance(bean)));
    }
}

File: android-beans/src/test/kotlin/rocks/frieler/android/beans/scopes/singleton/SingletonScopedFactoryBeanHandlerTest.kt
package rocks.frieler.android.beans.scopes.singleton

import assertk.assertThat
import assertk.assertions.isEqualTo
import assertk.assertions.isSameAs
import assertk.assertions.isSameInstanceAs
import assertk.assertions.isTrue
import org.junit.jupiter.api.Test
import org.mockito.Mockito.times
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import rocks.frieler.android.beans.BeansProvider

class SingletonScopedFactoryBeanHandlerTest {

    private val singletonScopedFactoryBeanHandler = SingletonScopedFactoryBeanHandler()

    private val scopedFactoryBean: SingletonScopedFactoryBean<Any> = mock()

    @Test
    fun `name is singleton scope`() {
        assertThat(singletonScopedFactoryBeanHandler.name).isEqualTo(SingletonScopedFactoryBeanHandler.SINGLETON_SCOPE)
    }

    @Test
    fun `singleton scope is always active`() {
        assertThat(singletonScopedFactoryBeanHandler.isActive).isTrue()
    }

    @Test
    fun `getBean() lets the FactoryBean produce the bean on first invocation`() {
        val dependencies: BeansProvider = mock()
        val bean = Any()
        whenever(scopedFactoryBean.produceBean(dependencies)).thenReturn(bean)

        val beanInstance = singletonScopedFactoryBeanHandler.getBean("bean", scopedFactoryBean, dependencies)

        assertThat(beanInstance).isSameInstanceAs(bean)
    }

    @Test
    fun `getBean() reuses singleton bean instance on every invocation`() {
        val dependencies: BeansProvider = mock()
        val bean = Any()
        whenever(scopedFactoryBean.produceBean(dependencies)).thenReturn(bean)

        val firstBeanInstance = singletonScopedFactoryBeanHandler.getBean("bean", scopedFactoryBean, dependencies)
        val secondBeanInstance = singletonScopedFactoryBeanHandler.getBean("bean", scopedFactoryBean, dependencies)

        verify(scopedFactoryBean, times(1)).produceBean(dependencies)
        assertThat(firstBeanInstance).isSameInstanceAs(bean)
        assertThat(secondBeanInstance).isSameInstanceAs(bean)
    }
}
